humility-core/src/hubris.rs

```diff
@@ -5362,6 +5475,24 @@ impl<'a> HubrisType<'a> {
             }
         }
     }
+
+    pub fn differs(&self, hubris: &HubrisArchive, rhs: &Self) -> Result<bool> {
```

Comment:
Architecture question:

It looks like `HubrisType::differs` is only ever called after using `HubrisArchive::lookup_type` to go from `HubrisGoff -> HubrisType`.

Should we instead write the function `HubrisArchive::differs(a: HubrisGoff, b: HubrisGoff) -> Result<bool>`?

Here's a diff of what that looks like:
```diff
diff --git a/humility-core/src/hubris.rs b/humility-core/src/hubris.rs
index dd807c0..d7e2dfd 100644
--- a/humility-core/src/hubris.rs
+++ b/humility-core/src/hubris.rs
@@ -3045,21 +3045,21 @@ impl HubrisArchive {
             Ok(goffs)
         } else {
             for goff in goffs {
-                cmp.push((goff, self.lookup_type(goff)?));
+                cmp.push(goff);
             }
 
             for i in 0..cmp.len() {
                 let mut dup = false;
 
                 for j in i + 1..cmp.len() {
-                    if !cmp[i].1.differs(self, &cmp[j].1)? {
+                    if !self.differs(cmp[i], cmp[j])? {
                         dup = true;
                         break;
                     }
                 }
 
                 if !dup {
-                    out.push(cmp[i].0);
+                    out.push(cmp[i]);
                 }
             }
             Ok(out)
@@ -4963,6 +4963,30 @@ impl HubrisArchive {
     pub fn read_cmpa(&self) -> Result<Option<Vec<u8>>> {
         self.read_file("img/CMPA.bin")
     }
+
+    pub fn differs(&self, a: HubrisGoff, b: HubrisGoff) -> Result<bool> {
+        let lhs = self.lookup_type(a)?;
+        let rhs = self.lookup_type(b)?;
+        match (lhs, rhs) {
+            (HubrisType::Base(lt), HubrisType::Base(rt)) => {
+                Ok(lt.encoding != rt.encoding || lt.size != rt.size)
+            }
+            (HubrisType::Struct(lt), HubrisType::Struct(rt)) => {
+                lt.differs(self, rt)
+            }
+            (HubrisType::Enum(lt), HubrisType::Enum(rt)) => {
+                lt.differs(self, rt)
+            }
+            (HubrisType::Array(lt), HubrisType::Array(rt)) => {
+                lt.differs(self, rt)
+            }
+            (HubrisType::Union(lt), HubrisType::Union(rt)) => {
+                lt.differs(self, rt)
+            }
+            (HubrisType::Ptr(l), HubrisType::Ptr(r)) => self.differs(l, r),
+            _ => Ok(true),
+        }
+    }
 }
 
 #[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
@@ -5124,10 +5148,7 @@ impl HubrisStruct {
                     return Ok(true);
                 }
 
-                let l = hubris.lookup_type(lm.goff)?;
-                let r = hubris.lookup_type(rm.goff)?;
-
-                if l.differs(hubris, &r)? {
+                if hubris.differs(lm.goff, rm.goff)? {
                     return Ok(true);
                 }
             }
@@ -5154,10 +5175,7 @@ impl HubrisArray {
         if self.count != rhs.count {
             Ok(true)
         } else {
-            let l = hubris.lookup_type(self.goff)?;
-            let r = hubris.lookup_type(rhs.goff)?;
-
-            l.differs(hubris, &r)
+            hubris.differs(self.goff, rhs.goff)
         }
     }
 }
@@ -5323,10 +5341,7 @@ impl HubrisEnum {
 
                 match (lm.goff, rm.goff) {
                     (Some(lg), Some(rg)) => {
-                        let l = hubris.lookup_type(lg)?;
-                        let r = hubris.lookup_type(rg)?;
-
-                        if l.differs(hubris, &r)? {
+                        if hubris.differs(lg, rg)? {
                             return Ok(true);
                         }
                     }
@@ -5387,10 +5402,7 @@ impl HubrisUnion {
 
                 match (lm.goff, rm.goff) {
                     (Some(lg), Some(rg)) => {
-                        let l = hubris.lookup_type(lg)?;
-                        let r = hubris.lookup_type(rg)?;
-
-                        if l.differs(hubris, &r)? {
+                        if hubris.differs(lg, rg)? {
                             return Ok(true);
                         }
                     }
@@ -5475,24 +5487,6 @@ impl<'a> HubrisType<'a> {
             }
         }
     }
-
-    pub fn differs(&self, hubris: &HubrisArchive, rhs: &Self) -> Result<bool> {
-        match (self, rhs) {
-            (Self::Base(lt), Self::Base(rt)) => {
-                Ok(lt.encoding != rt.encoding || lt.size != rt.size)
-            }
-            (Self::Struct(lt), Self::Struct(rt)) => lt.differs(hubris, rt),
-            (Self::Enum(lt), Self::Enum(rt)) => lt.differs(hubris, rt),
-            (Self::Array(lt), Self::Array(rt)) => lt.differs(hubris, rt),
-            (Self::Union(lt), Self::Union(rt)) => lt.differs(hubris, rt),
-            (Self::Ptr(l), Self::Ptr(r)) => {
-                let lt = hubris.lookup_type(*l)?;
-                let rt = hubris.lookup_type(*r)?;
-                lt.differs(hubris, &rt)
-            }
-            _ => Ok(true),
-        }
-    }
 }
 
 impl<'a> From<&'a HubrisBasetype> for HubrisType<'a> {
```

It's a small LOC decrease, but I'm not 100% sold on this, because it does more work in `dedup` (calling `lookup_type` repeatedly).

Reply:
Yeah, dunno on that one. I feel I could go either way. It *does* feel slightly cleaner -- and may make for a more useful future interface...

Reply:
Ended up doing the refactor per the suggestion -- thank you!